### ISSUE TITLE
Reading from channel hangs if there's only STDERR output

### ISSUE THREAD

**Issue by ferki** [rkitover/net-ssh2#32]
I might have found a regression in 0.59_10 compared to 0.58. If a remote command only prints to STDERR, the following example code works with 0.58 to retrieve STDOUT and STDERR separately, but hangs while reading from STDOUT with 0.59_10:
```perl
use strict;
use warnings;
use 5.010;

use Net::SSH2;

my $ssh2 = Net::SSH2->new();
$ssh2->connect('myhost') or die $!;
say 'connected';
$ssh2->auth_publickey( 'user', 'my_pub_key', 'my_priv_key' );
say 'authenticated';

my $chan = $ssh2->channel();
$chan->exec('perl -E "say STDERR \"err\""');
say 'executed';

my $len;

say 'stdout:';
$len = $chan->read( my $out_buf, 2048 );
say $out_buf if $len;

say 'stderr:';
$len = $chan->read( my $err_buf, 2048, 1 );
say $err_buf if $len;

say 'exit code: ' . $chan->exit_status;
```

It may very well be the new intended behaviour, but I didn't find another way to retrieve STDOUT and STDERR separately so far based on the docs. In this case, please advise how to go about that.

**Comment by salva** [comment]
it seems both things happen:

* There is a bug on `Net::SSH2::Channel::read`: the underlaying `libssh2_channel_read` may return 0 with out it being an error and `read` was trying to hide that from the user (i.e. you), in order to mimic Perl built-ins behavior, checking whether the channel was at EOF or otherwise, just retrying the operation. The problem is that on SSH the EOF flag is shared between stdout and stderr.

* your code is wrong: It is as in the Net::OpenSSH case, you have to set non blocking mode and read from both strings in parallel. But let me make my mind about that because using non-blocking mode in Net::SSH2/libssh2 is quite difficult and full of traps. I think I would be able to provide you a function to read from both streams in parallel.

**Comment by ferki** [comment]
> It is as in the Net::OpenSSH case, you have to set non blocking mode and read from both strings in parallel. 

Yeah, it looks like the Net::SSH2 parts were not updated in Rex yet.. :disappointed: I can take care of that. 

> I think I would be able to provide you a function to read from both streams in parallel.

I'll wait for your feedback about it first.

**Comment by salva** [comment]
I have released version [0.59_11](https://metacpan.org/release/SALVA/Net-SSH2-0.59_11) fixing that bug.

That version adds also support for the [read2](https://metacpan.org/pod/release/SALVA/Net-SSH2-0.59_11/lib/Net/SSH2/Channel.pm#read2-max_size) method which allows to read both `stdout` and `stderr`.

**Comment by ferki** [comment]
@salva: thank you! I tested it successfully (by building from source). I'll convert related rex code.
